apps/web/app/utils/extract-from-formdata.ts

```diff
@@ -13,3 +13,66 @@ export function extract<T extends ZodTypeAny>(
   }
   return parsed.data;
 }
+
+export type ParseFromDataRes<T extends ZodObject<any>> =
+  | {
+      success: true;
+      data: z.infer<T>;
+      error: null;
+    }
+  | {
+      success: false;
+      data: null;
+      error: ZodError;
+    };
+
+export function parseFormData<T extends ZodObject<any>>(
+  formData: FormData,
+  parser: T
+): ParseFromDataRes<T> {
+  const keys = Object.keys(parser.shape);
+
+  const data: Record<string, any> = {};
+
+  for (const key of keys) {
+    data[key] = formData.get(key);
+  }
+
+  const parsed = parser.safeParse(data);
+
+  if (!parsed.success) {
+    return { success: false, data: null, error: parsed.error };
+  }
+
+  return { success: true, data: parsed.data, error: null };
+}
+
+export type WithFormData = {
+  formData: () => Promise<FormData>;
+};
+
+export async function extractFromFormData<T extends ZodObject<any>>(
+  request: WithFormData,
+  schema: T,
+  error = badRequest("Error parsing body")
```

Comment:
I believe the `throw` in here can be a little deceiving, I like better the pattern of handling the error in the caller but that's just personal preference.

The only thing I do think we can improve is returning the actual errors, as I commented in the freeze page, otherwise we lose client info